packages/react-dom/src/__tests__/refs-destruction-test.js

```diff
@@ -46,77 +54,96 @@ describe('refs-destruction', () => {
         } else {
           return (
             <div>
-              <div ref={this.theInnerDivRef} />
-              <ClassComponent ref={this.theInnerClassComponentRef} />
+              <div ref={theInnerDivRef} />
+              <ClassComponent ref={theInnerClassComponentRef} />
             </div>
           );
         }
       }
     };
   });
 
-  it('should remove refs when destroying the parent', () => {
+  afterEach(() => {
+    theInnerClassComponentRef = null;
+    theInnerDivRef = null;
+  });
+
+  it('should remove refs when destroying the parent', async () => {
     const container = document.createElement('div');
-    const testInstance = ReactDOM.render(<TestComponent />, container);
+    const root = ReactDOMClient.createRoot(container);
+    await act(async () => {
+      root.render(<TestComponent />);
+    });
 
-    expect(
-      ReactTestUtils.isDOMComponent(testInstance.theInnerDivRef.current),
-    ).toBe(true);
-    expect(testInstance.theInnerClassComponentRef.current).toBeTruthy();
+    expect(ReactTestUtils.isDOMComponent(theInnerDivRef.current)).toBe(true);
+    expect(theInnerClassComponentRef.current).toBeTruthy();
 
-    ReactDOM.unmountComponentAtNode(container);
+    root.unmount();
```

Comment:
I'm surprised this works, I thought unmount is auto-batched like render.

Reply:
It's not, we wrap unmount in `flushSync`. If there was a `render` before `unmount` that hasn't flushed yet, then it would be batched with the sync flush for unmount: https://github.com/facebook/react/blob/f498aa299292d4d1e999f66d1c769440ad10d57c/packages/react-dom/src/client/ReactDOMRoot.js#L155